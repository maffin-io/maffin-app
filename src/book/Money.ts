import * as currencies from '@dinero.js/currencies';
import type { Currency as DineroCurrency } from 'dinero.js';
import * as djs from 'dinero.js';

import { toAmountWithScale, moneyToString } from '@/helpers/number';

export default class Money {
  private _raw: djs.Dinero<number>;

  /**
   * A wrapper class for dinero object to reduce verbosity and overhead. The constructor
   * checks for how many decimals the amount has and calculates the scale automatically.
   *
   * @param amount - The amount to be considered in the money object. Can be a float
   * @param currency - The currency
   * @param scale - Number representing custom scale. Default is 0, if passed differently, that one
   * is taken into account
   */
  constructor(amount: number, currency: string, scale?: number) {
    const dineroCurrency = toCurrency(currency);

    if (scale === undefined) {
      const { amount: newAmount, scale: newScale } = toAmountWithScale(amount);
      this._raw = djs.dinero({
        currency: dineroCurrency,
        amount: newAmount,
        scale: newScale,
      });
    } else {
      this._raw = djs.dinero({
        currency: dineroCurrency,
        amount,
        scale,
      });
    }
  }

  get currency(): string {
    return djs.toSnapshot(this._raw).currency.code;
  }

  get raw(): djs.Dinero<number> {
    return this._raw;
  }

  /**
   * Represents the money object as a string with the currency standardizing
   * the scale to 2 as a default. If the scale is passed then it uses that.
   */
  toString(scale = 2): string {
    return djs.toDecimal(
      djs.transformScale(this._raw, scale),
      ({ value, currency }) => `${value} ${currency.code}`,
    );
  }

  /**
   * Represents the money object as a string with the currency standardizing
   * the scale to 2 as a default. If the scale is passed then it uses that.
   *
   * The result is a localized string with the currency as a symbol
   */
  format(scale = 2): string {
    return djs.toDecimal(
      djs.transformScale(this._raw, scale),
      ({ value, currency }) => moneyToString(Number(value), currency.code),
    );
  }

  toNumber(): number {
    return parseFloat(djs.toDecimal(this._raw));
  }

  isNegative(): boolean {
    return djs.isNegative(this._raw);
  }

  isZero(): boolean {
    return djs.isZero(this._raw);
  }

  /**
   * Adds b to the current object and returns a new one with the result
   *
   * @param b - Money to be added
   * @returns - Money with result
   */
  add(b: Money): Money {
    const result = djs.toSnapshot(djs.add(b.raw, this.raw));
    return new Money(result.amount, this.currency, result.scale);
  }

  /**
   * Subtracts b to the current object and returns a new one with the result
   *
   * @param b - Money to be subtracted
   * @returns - Money with result
   */
  subtract(b: Money): Money {
    const result = djs.toSnapshot(djs.subtract(this.raw, b.raw));
    return new Money(result.amount, this.currency, result.scale);
  }

  /**
   * Multiplies b to the current object and returns a new one with the result
   * Initially, we try to multiply without transforming b but sometimes dinerojs
   * returns that it is an invalid amount (I have no idea why). If it fails, then
   * we transform b to `amountWithScale` and multiply with that.
   *
   * @param b - amount to be multiplied to in scale 0
   * @returns - Money with result
   */
  multiply(b: number): Money {
    let result;

    try {
      result = djs.multiply(this.raw, b);
    } catch {
      result = djs.multiply(this.raw, toAmountWithScale(b));
    }

    const snapshot = djs.toSnapshot(result);
    return new Money(snapshot.amount, this.currency, snapshot.scale);
  }

  /**
   * Checks for equality with b by comparing amount and currency
   *
   * @returns - boolean
   */
  equals(b: Money): boolean {
    return djs.equal(this._raw, b._raw);
  }

  /**
   * Converts a Money object to the currency specified
   *
   * @param to - Money object containing the rate to
   */
  convert(
    to: string,
    rate: number,
  ): Money {
    const rates = {
      [to]: {
        ...toAmountWithScale(rate),
      },
    };
    const currency = toCurrency(to);

    try {
      // @ts-ignore
      const result = djs.toSnapshot(djs.convert(this.raw, currency, rates));
      return new Money(result.amount, currency.code, result.scale);
    } catch (e) {
      throw new Error(`Error converting ${this.format()} to ${to} with rates ${JSON.stringify(rates)}: ${e}`);
    }
  }
}

function toCurrency(currency: string): DineroCurrency<number> {
  // @ts-ignore
  let dineroCurrency = currencies[currency];
  if (!dineroCurrency) {
    dineroCurrency = {
      code: currency,
      base: 10,
      exponent: 2,
    };
  }
  return dineroCurrency;
}
