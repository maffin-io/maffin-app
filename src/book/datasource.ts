import initSqlJs from 'sql.js';
import { DataSource } from 'typeorm';
import { useSWRConfig } from 'swr';

import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from './entities';

export async function initDB(database: Uint8Array): Promise<DataSource> {
  await initSqlJs({
    // Required to load the wasm binary asynchronously. Of course, you can host it wherever you want
    locateFile: file => `https://sql.js.org/dist/${file}`,
  });

  let datasource: DataSource;

  if (database.length !== 0) {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      database,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
  } else {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
    await createEmptyBook();
  }

  return datasource;
}

// TODO: Need to mutate the accounts SWR key so Root is shown!
async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}
