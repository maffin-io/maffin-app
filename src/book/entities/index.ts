export { default as Account } from './Account';
export { default as BankConfig } from './BankConfig';
export { default as Commodity } from './Commodity';
export { default as Price } from './Price';
export { default as Split } from './Split';
export { default as Transaction } from './Transaction';
export { default as Book } from './Book';
export { default as BaseEntity } from './BaseEntity';
