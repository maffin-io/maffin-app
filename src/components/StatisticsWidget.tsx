import React from 'react';

export type StatisticsWidgetProps = {
  title: string,
  stats: string,
  description: string,
  className?: string,
  statsTextClass?: string,
};

export default function StatisticsWidget({
  title,
  stats,
  description,
  className = '',
  statsTextClass = '',
}: StatisticsWidgetProps): JSX.Element {
  return (
    <div className={`${className} bg-gunmetal-700 rounded-sm p-6`}>
      <p>
        {title}
      </p>
      <p className={`${statsTextClass} text-2xl font-semibold my-3`}>{stats}</p>

      <span className="text-sm">
        {description}
      </span>
    </div>
  );
}
