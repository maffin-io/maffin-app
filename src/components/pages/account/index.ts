export { default as TotalLineChart } from './TotalLineChart';
export { default as SplitsHistogram } from './SplitsHistogram';
export { default as TransactionsTable } from './TransactionsTable';
export { default as Header } from './Header';
export { default as AccountInfo } from './AccountInfo';
export { default as InvestmentInfo } from './InvestmentInfo';
export { default as AssetInfo } from './AssetInfo';
