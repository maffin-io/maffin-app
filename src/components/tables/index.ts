export { default as InvestmentsTable } from './InvestmentsTable';
export { default as AccountsTable } from './AccountsTable';
export { default as TransactionsTable } from './TransactionsTable';
export { default as PricesTable } from './PricesTable';
