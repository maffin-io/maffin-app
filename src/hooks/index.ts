export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
