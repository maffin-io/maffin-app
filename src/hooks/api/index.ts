import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
