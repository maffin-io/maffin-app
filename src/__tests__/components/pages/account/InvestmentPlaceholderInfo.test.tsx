import React from 'react';
import { render, screen } from '@testing-library/react';
import type { UseQueryResult } from '@tanstack/react-query';

import { Account, Commodity } from '@/book/entities';
import StatisticsWidget from '@/components/StatisticsWidget';
import {
  WeightsChart,
  DividendChart,
} from '@/components/charts';
import { InvestmentsTable } from '@/components/tables';
import Money from '@/book/Money';
import { InvestmentAccount } from '@/book/models';
import { InvestmentPlaceholderInfo } from '@/components/pages/account';
import * as apiHook from '@/hooks/api';

jest.mock('@/hooks/api', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/api'),
}));

jest.mock('@/components/charts/WeightsChart', () => jest.fn(
  () => <div data-testid="WeightsChart" />,
));
const WeightsChartMock = WeightsChart as jest.MockedFunction<typeof WeightsChart>;

jest.mock('@/components/StatisticsWidget', () => jest.fn(
  () => <div data-testid="StatisticsWidget" />,
));

jest.mock('@/components/charts/DividendChart', () => jest.fn(
  () => <div data-testid="DividendChart" />,
));

jest.mock('@/components/tables/InvestmentsTable', () => jest.fn(
  () => <div data-testid="InvestmentsTable" />,
));

jest.mock('@/components/Loading', () => jest.fn(
  () => <div data-testid="Loading" />,
));

describe('InvestmentPlaceholderInfo', () => {
  beforeEach(() => {
    jest.spyOn(apiHook, 'useInvestments').mockReturnValue({ data: undefined } as UseQueryResult<InvestmentAccount[]>);
    jest.spyOn(apiHook, 'useMainCurrency').mockReturnValue({ data: undefined } as UseQueryResult<Commodity>);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('shows loading when loading data', async () => {
    jest.spyOn(apiHook, 'useInvestments').mockReturnValue({ isPending: true } as UseQueryResult<InvestmentAccount[]>);
    render(<InvestmentPlaceholderInfo account={{ guid: 'guid' } as Account} />);

    await screen.findByTestId('Loading');
  });

  it('renders with no investments', async () => {
    jest.spyOn(apiHook, 'useInvestments').mockReturnValue({ data: [] as InvestmentAccount[] } as UseQueryResult<InvestmentAccount[]>);
    render(<InvestmentPlaceholderInfo account={{ guid: 'guid' } as Account} />);
    await screen.findByText('You have no investments yet!');
  });

  it('renders with data', async () => {
    const mainCurrency = {
      mnemonic: 'EUR',
      guid: 'eur',
      namespace: 'CURRENCY',
    } as Commodity;

    const investment1 = {
      account: { guid: 'guid1', name: 'Investment' },
      quantity: new Money(10, 'TICKER'),
      valueInCurrency: new Money(5, 'EUR'),
      costInCurrency: new Money(4, 'EUR'),
      realizedProfitInCurrency: new Money(10, 'EUR'),
      realizedDividendsInCurrency: new Money(5, 'EUR'),
    } as InvestmentAccount;

    jest.spyOn(apiHook, 'useInvestments').mockReturnValueOnce({ data: [investment1] } as UseQueryResult<InvestmentAccount[]>);
    jest.spyOn(apiHook, 'useMainCurrency').mockReturnValue({ data: mainCurrency } as UseQueryResult<Commodity>);

    const { container } = render(<InvestmentPlaceholderInfo account={{ childrenIds: ['guid1'] } as Account} />);

    await screen.findByTestId('InvestmentsTable');
    expect(WeightsChart).toHaveBeenCalledWith(
      {
        accounts: [investment1.account.guid],
        totalValue: expect.any(Money),
      },
      {},
    );
    expect(WeightsChartMock.mock.calls[0][0].totalValue.toString()).toEqual('5 EUR');
    expect(StatisticsWidget).toBeCalledTimes(3);
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      1,
      {
        className: 'mr-2',
        title: 'Value/Cost',
        stats: '€5.00',
        description: '€4.00 total invested',
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      2,
      {
        className: 'mr-2',
        title: 'Unrealized Profit',
        stats: '€1.00 (25%)',
        description: '€6.00 (150%) with dividends',
        statsTextClass: 'amount-positive',
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      3,
      {
        className: 'mr-2',
        title: 'Realized',
        stats: '€15.00',
        description: '€5.00 from dividends',
        statsTextClass: 'amount-positive',
      },
      {},
    );
    expect(DividendChart).toHaveBeenLastCalledWith(
      {
        accounts: [investment1.account.guid],
      },
      {},
    );
    expect(InvestmentsTable).toHaveBeenLastCalledWith(
      {
        accounts: [investment1.account.guid],
      },
      {},
    );
    expect(container).toMatchSnapshot();
  });

  it('renders with different mainCurrency', async () => {
    const mainCurrency = {
      mnemonic: 'USD',
      guid: 'usd',
      namespace: 'CURRENCY',
    } as Commodity;

    const investment1 = {
      account: { guid: 'guid1', name: 'Investment' },
      quantity: new Money(10, 'TICKER'),
      valueInCurrency: new Money(5, 'USD'),
      costInCurrency: new Money(4, 'USD'),
      realizedProfitInCurrency: new Money(10, 'USD'),
      realizedDividendsInCurrency: new Money(5, 'USD'),
    } as InvestmentAccount;

    jest.spyOn(apiHook, 'useInvestments').mockReturnValueOnce({ data: [investment1] } as UseQueryResult<InvestmentAccount[]>);
    jest.spyOn(apiHook, 'useMainCurrency').mockReturnValue({ data: mainCurrency } as UseQueryResult<Commodity>);

    render(<InvestmentPlaceholderInfo account={{ childrenIds: ['guid1'] } as Account} />);

    await screen.findByTestId('InvestmentsTable');
    expect(WeightsChartMock.mock.calls[0][0].totalValue.toString()).toEqual('5 USD');
    expect(StatisticsWidget).toBeCalledTimes(3);
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      1,
      {
        className: 'mr-2',
        title: 'Value/Cost',
        stats: '$5.00',
        description: '$4.00 total invested',
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      2,
      {
        className: 'mr-2',
        title: 'Unrealized Profit',
        stats: '$1.00 (25%)',
        description: '$6.00 (150%) with dividends',
        statsTextClass: 'amount-positive',
      },
      {},
    );
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      3,
      {
        className: 'mr-2',
        title: 'Realized',
        stats: '$15.00',
        description: '$5.00 from dividends',
        statsTextClass: 'amount-positive',
      },
      {},
    );
  });
});
