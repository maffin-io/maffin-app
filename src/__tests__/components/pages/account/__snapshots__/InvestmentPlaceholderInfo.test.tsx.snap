// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvestmentPlaceholderInfo renders with data 1`] = `
<div>
  <div
    class="grid md:grid-cols-12 auto-cols-fr mt-4"
  >
    <div
      class="md:col-span-4 mt-2"
    >
      <div
        data-testid="WeightsChart"
      />
    </div>
    <div
      class="md:col-span-8"
    >
      <div
        class="grid md:grid-cols-12 auto-cols-fr"
      >
        <div
          class="md:col-span-4"
        >
          <div
            data-testid="StatisticsWidget"
          />
        </div>
        <div
          class="md:col-span-4"
        >
          <div
            data-testid="StatisticsWidget"
          />
        </div>
        <div
          class="md:col-span-4"
        >
          <div
            data-testid="StatisticsWidget"
          />
        </div>
      </div>
      <div
        class="card"
      >
        <div
          data-testid="DividendChart"
        />
      </div>
    </div>
  </div>
  <div
    class="py-4"
  >
    <div
      data-testid="InvestmentsTable"
    />
  </div>
</div>
`;
